omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_9/ConceptReferenceTermSearchHandler1_9.java

```diff
@@ -75,84 +77,119 @@ public PageableResult search(RequestContext context) throws ResponseException {
+			if (singleConceptSource != null) {
+				conceptSources.add(singleConceptSource);
+			}
+			else {
+				List<ConceptSource> allConceptSources = conceptService.getAllConceptSources();
+				if (allConceptSources.size() == 0) {
+					return new EmptySearchResult();
+				}
+				else {
+					for (ConceptSource conceptSource : allConceptSources) {
+						if (conceptSource.getName().contains(source)) {
+							conceptSources.add(conceptSource);
+						}
+					}
+				}
+			}
+			return searchStartsWith(context, code, name, conceptSources);
+		}
+		else {
+			throw new IllegalStateException("\"" + searchType + "\" doesn't fit searchType param requirements. Search type must be " + SEARCH_TYPE_EQUALS + " or " + SEARCH_TYPE_STARTS_WITH + ".");
 		}
 	}
 
-	private PageableResult searchStartsWith(RequestContext context, String code, String name, ConceptSource conceptSource){
+	private PageableResult searchStartsWith(RequestContext context, String code, String name, List<ConceptSource> conceptSources){
 		/**
 		 * org.openmrs.api.ConceptService#getConceptReferenceTerms(java.lang.String, org.openmrs.ConceptSource, java.lang.Integer, java.lang.Integer, boolean)
 		 * returns all ConceptReferenceTerm objects which code or name match query, so results need to be filtered
 		 */
 		List<ConceptReferenceTerm> resultTerms = new ArrayList<ConceptReferenceTerm>();
-		if(code != null){
-			List<ConceptReferenceTerm> codeQueryTerms = conceptService.getConceptReferenceTerms(code, conceptSource, context.getStartIndex(),
-					context.getLimit(), context.getIncludeAll());
-			for(ConceptReferenceTerm term : codeQueryTerms){
-				if(termMatchesCodeAndName(term, code, name)){
-					resultTerms.add(term);
+
+		if ((code != null && code.length() > 0) || (name != null && name.length() > 0)) {
+
+			Set<ConceptReferenceTerm> referenceTermsByCode = new HashSet<ConceptReferenceTerm>();
+			Set<ConceptReferenceTerm> referenceTermsByName = new HashSet<ConceptReferenceTerm>();
+
+			if (code != null && code.length() > 0) {
+				for (ConceptSource conceptSource : conceptSources) {
+					referenceTermsByCode.addAll(conceptService.getConceptReferenceTerms(code, conceptSource, context.getStartIndex(), context.getLimit(), context.getIncludeAll()));
 				}
 			}
-		}
-		if(name != null){
-			List<ConceptReferenceTerm> nameQueryTerms = conceptService.getConceptReferenceTerms(name, conceptSource, context.getStartIndex(),
-					context.getLimit(), context.getIncludeAll());
-			for(ConceptReferenceTerm term : nameQueryTerms){
-				if(termMatchesCodeAndName(term, code, name)){
-					resultTerms.add(term);
+			else {
+				resultTerms.addAll(referenceTermsByName);
+			}
+
+			if (name != null && name.length() > 0) {
+				for (ConceptSource conceptSource : conceptSources) {
+					referenceTermsByName.addAll(conceptService.getConceptReferenceTerms(name, conceptSource, context.getStartIndex(), context.getLimit(), context.getIncludeAll()));

```

Comment:
You seem to be calling the same method for names and codes: conceptService.getConceptReferenceTerms

It doesn't seem right...